src/main/java/org/opensearch/knn/plugin/KNNPlugin.java

```diff
@@ -281,10 +282,14 @@ public Optional<EngineFactory> getEngineFactory(IndexSettings indexSettings) {
         return Optional.empty();
     }
 
+    /**
+     * Get KNN codec service.
+     * We create {@link RemoteIndexBuilder} per-index in order to consume the {@link IndexSettings} from the {@link org.apache.lucene.codecs.KnnVectorsWriter} implementation.
+     */
     @Override
     public Optional<CodecServiceFactory> getCustomCodecServiceFactory(IndexSettings indexSettings) {
         if (indexSettings.getValue(KNNSettings.IS_KNN_INDEX_SETTING)) {
-            return Optional.of(KNNCodecService::new);
+            return Optional.of((config) -> new KNNCodecService(config, new RemoteIndexBuilder(repositoriesServiceSupplier, indexSettings)));
```

Comment:
I like this idea of passing index settings during codec creation to RemoteIndex builder rather than just using index name and getting index settings from cluster state from any part of the code. 

This is good thinking.